fendermint/eth/api/src/conv/from_tm.rs

```diff
@@ -342,33 +351,60 @@ fn maybe_contract_address(deliver_tx: &DeliverTx) -> Option<EthAddress> {
     fendermint_rpc::response::decode_fevm_create(deliver_tx)
         .ok()
         .map(|cr| {
-            // We can return `cr.eth_address` here and that's an address usable for calling the contract.
-            // However, the events are reported using the actor ID, so unless we want to translate those
-            // to ETH addresses on the fly, we can return a masked address here and see if that works.
-            EthAddress::from_id(cr.actor_id)
+            // We can return either `cr.actor_id` as a masked address,
+            // or `cr.eth_address`. Both addresses are usable for calling the contract.
+            // However, the masked ID doesn't work with some of the Ethereum tooling which check some hash properties.
+            // We also have to make sure to use the same kind of address that we do in the filtering and event logs,
+            // otherwise the two doesn't align and it makes the API difficult to use. It's impossible(?) to find out
+            // the actor ID just using the Ethereum API, so best use the same.
+            // EthAddress::from_id(cr.actor_id)
+            cr.eth_address
         })
 }
 
-pub fn to_logs(
+/// Turn Events into Ethereum logs.
+///
+/// We need to turn Actor IDs into Ethereum addresses because that's what the tooling expects.
+pub async fn to_logs<C>(
+    client: &FendermintClient<C>,
     events: &[abci::Event],
     block_hash: et::H256,
     block_number: et::U64,
     transaction_hash: et::H256,
     transaction_index: et::U64,
     log_index_start: usize,
-) -> anyhow::Result<Vec<et::Log>> {
+) -> anyhow::Result<Vec<et::Log>>
+where
+    C: Client + Sync + Send,
+{
     let mut logs = Vec::new();
     for (idx, event) in events.iter().enumerate() {
-        // TODO: Lotus looks up an Ethereum address based on the actor ID:
+        // Lotus looks up an Ethereum address based on the actor ID:
         // https://github.com/filecoin-project/lotus/blob/6cc506f5cf751215be6badc94a960251c6453202/node/impl/full/eth.go#L1987
-        let address = event
+        let actor_id = event
             .attributes
             .iter()
             .find(|a| a.key == "emitter")
             .and_then(|a| a.value.parse::<u64>().ok())
-            .map(EthAddress::from_id)
-            .map(|a| et::H160::from_slice(&a.0))
-            .unwrap_or_default();
+            .ok_or_else(|| anyhow!("cannot find the 'emitter' key"))?;
+
+        // TODO: Cache the emitters to save API roundtrips.
```

Comment:
This can potentially be removed once we fix the return inconsistency between Eth and Fendermint transactions returned, right?

Reply:
No, it can't, because the `emitter` in the FVM [StampedEvent](https://github.com/filecoin-project/ref-fvm/blob/master/shared/src/event/mod.rs#L15) is an `ActorID`, and we can't possibly know what to turn it into. Ostensibly we could emit a "delegated emitter" too, and then we could filter on that one, or we do what we do now to translate on the fly for Ethereum.

Reply:
The fix for the transaction hash is similar in that it will rely on adding further indexed fields, so we can explore this idea as well.